### ISSUE TITLE
The performance of pre-trained model

### ISSUE THREAD

**Issue by shuni1001** [yikang-li/FactorizableNet#20]
I ran your code with your trained-model for VG-MSDN dataset, but I faced huge gap between your result and mine.

In the ReadMe document, the performance of  provided (pre-trained for VG-MSDN) model is  mentioned Recall@50 12.984% and Recall@100 16.506%. 

However, I ran your code with provided model, I got below result.
![Screen Shot 2019-04-14 at 9 20 56 PM](https://user-images.githubusercontent.com/7753796/56092778-3590e000-5efb-11e9-987b-0638c8b67d17.png)

I download visual genome dataset from official website, there are two zip files images.zip and images2.zip. 
After download them,  I merged two directory (images, images2) into single directory, then create symbolic link as mentioned your ReadMe File.

Could you give me some hint for what I missed?

**Comment by Freya-Cui** [comment]
Hello, could you tell me how much memory did your GPU use while testing with the trained model?
I have got an error `CUDA out of memory` while testing.

**Comment by shuni1001** [comment]
Hello, my gpus are titan xp which has 12Gb memory for each gpu.
The exp was run on single gpu.

**Comment by Freya-Cui** [comment]
Thank you for your reply! :)
I have solved the problem by adding `with torch.no_grad()` before inference.
By the way, I got the same testing result as you.

**Comment by sandeep-ipk** [comment]
Hey @shuni1001, did you re-evaluate the pretrained model? Did you find any improvement in the results?

**Comment by sandeep-ipk** [comment]
@shuni1001 , @Freya-Cui , I think you are not using "use_region" option in the model options file:
(VG-MSDN.yaml). It is set to false by default.
Since you are not using this external (phrases for regions) knowledge base, I think we are not getting the results as reported in README.

**Comment by yikang-li** [comment]
Hi @sandeep-ipk  we didn't use the external captions, like caption regions, during the training in this job. 

The provided pre-trained model is trained using warm-up training. We trained the model for several epochs with learning rate decay, then reset the learning rate and train again. It will help to get a better result. I have provided the information in the last block of the "Training" section in README.

**Comment by muraliadithya** [comment]
@Freya-Cui Where did you add the 'with torch.no_grad()' ? Could you tell me the file name and the line number?

**Comment by emanuelazfernandes** [comment]
@muraliadithya and anyone who might come across this:

TL;DR: add it before this line [train_FN.py#L276](https://github.com/yikang-li/FactorizableNet/blob/master/train_FN.py#L276), and indent everything until return at [#L291](https://github.com/yikang-li/FactorizableNet/blob/master/train_FN.py#L291), like this:

```python
    if args.evaluate:
        with torch.no_grad():
            recall, result = model.module.engines.test(test_loader, model, top_Ns,
                                                nms=args.nms,
                                                triplet_nms=args.triplet_nms,
                                                use_gt_boxes=args.use_gt_boxes)
            print('======= Testing Result =======')
            for idx, top_N in enumerate(top_Ns):
                print('Top-%d Recall'
                      '\t[Pred]: %2.3f%%'
                      '\t[Phr]: %2.3f%%'
                      '\t[Rel]: %2.3f%%' % (
                        top_N, float(recall[2][idx]) * 100,
                        float(recall[1][idx]) * 100,
                        float(recall[0][idx]) * 100))
            print('============ Done ============')
            save_results(result, None, options['logs']['dir_logs'], is_testing=True)
        return
```

In PyTorch, gradients are required only during training for the backpropagation step, so during inference (testing), you're just performing a forward pass and gradients are not needed. However, by default, PyTorch keeps track of operations to compute gradients, even during inference, which consumes unnecessary memory.
A simple way to avoid this and free that useless extra memory, is to use the context manager with `torch.no_grad()`, so that locally that piece of code which contains the inference functionality doesn't use the gradients.

For the Visual Genome MSDN trained model, the difference in GPU memory usage is from ~11 GB to ~8 GB. I've only tested on my local hardware, a 2080 Ti which has 11 GB and I was running into the issue, but we can assume the authors had no trouble with the good ol' Titan which had 12 GB.
